docker/compose-controller-spark-sql.yaml

```diff
@@ -36,6 +36,7 @@ services:
     build:
       context: ..
     container_name: pipeline-controller
+    restart: on-failure
```

Comment:
What is the rationale behind this? I mean is there a particular scenario that this fails and we expect it to succeed on restart? If this is because of Metastore table creation on start, should we make this wait for the thriftserver service, e.g., using `depends_on`?

Reply:
I did try `depends-on` and observed that the pipeline controller does not wait for thriftserver service to get started and rather it checks for its `ready` stage. At this point, thriftserver is not ready to accept connection on 10000 port and pipeline controller on its start tries to establish this connection and fails. And hence, I added this feature so that it can restart and try to connect again.

I also tried healthcheck and `service_healthy` option but it did not work and that requires hitting service continuously within certain time period.

I can still go ahead and add the dependency.

Here is some useful documentation and examples:
https://docs.docker.com/compose/compose-file/05-services/#depends_on
https://stackoverflow.com/questions/31746182/docker-compose-wait-for-container-x-before-starting-y

Reply:
I see, so why didn't `healthcheck` work? I mean couldn't you try connecting to port 10000 and retry until it succeed? If that is not doable, is it possible to move the check to our Java code, i.e., when trying to connect to the Metastore, if it fails, it retries again, e.g., several times in intervals of few seconds? This is not a big deal but it is a little bit ugly to expect failure and restart.

Reply:
this is fixed now with healthcheck